contracts/specs/NFTSwapper/NFTSwapperSpec.md

```diff
@@ -0,0 +1,163 @@
+
+Pros:
+* Simplicity
+
+Cons:
+* Values all NFTs equally, worst for maintaining non-fungibility.
+
+### Output NFT age based curve
+
+Prices lower `outputTokenId` at a higher value. 
+
+```solidity
+struct PriceCurveParams{
+    uint256 floor;
+    uint256 slope;
+}
+
+PriceCurveParams immutable priceCurveParams;
+
+function swapPrice(uint256 outputTokenId) view returns (uint256) {
+    uint256 outputAge = nft.totalSupply() - outputTokenId;
+    return priceCurveParams.floor + outputAge * priceCurveParams.slope;
+}
+```
+
+![](img/output-age-based-price-curve.png)
+
+Pros:
+* Values output NFTs distinctly, where lower ID's are more expensive to swap for.
+
+Cons: 
+* Treats all `inputTokenIds` as being valued equally (i.e treats as fungible)
+* Requires the ERC721 is `ERC721Enumerable` to support the `totalSupply()` 
+* Time variance, the `swapPrice` gets more expensive every day (could argue this is a pro)
+
+### Input and Output NFT age difference based curve
+
+Price based on the difference between `inputTokenId` and `outputTokenId` (or age difference). 
+
+```solidity
+struct PriceCurveParams{
+    uint256 floor;
+    uint256 slope;
+}
+
+PriceCurveParams immutable priceCurveParams;
+
+function swapPrice(uint256 inputTokenId, uint256 outputTokenId) pure returns (uint256) {
+    (, uint256 ageDiff) = inputTokenId.trySub(outputTokenId); // Clamps to 0
+    return priceCurveParams.floor + ageDiff * priceCurveParams.slope;
+}
+```
+
+![](img/age-difference-price-curve.png)
+
+Pros:
+* Values output NFTs distinctly, where lower ID's are more expensive to swap for.
+* Values input NFTs distinctly, where lower ID's are less expensive to swap from.
+* Time invariance, the `swapPrice` today will be the same as tomorrow for the same swap.
+
+Cons:
+* Most complex in terms of user understanding (?) 
+
+## Approval Considerations
+
+As mentioned above, the `swapPool` must approve this contract to transfer any NFTs which are desired to be available to be swapped. The easiest option is to `setApprovalForAll` to make all NFTs held by the `swapPool` available for swapping. 
+
+An alternative to restrict the swappable NFTs to a subset of those held in the `swapPool` is to individually approve via the `approve` function. When using individual approvals, every swap will reduce the swappable number of NFTs by one. This means that batch approvals will likely need to continue to happen over time to "refill" the swappable NFTs. 
+
+NFTs which are approved for swapping can still be used by the `swapPool` for other things, but note that if the `swapPool` is a treasury with some timelock (like in Nouns), and the timelock execution is trying to move an NFT which is also approved for swapping, then a malicious actor could block this execution by swapping for that NFT before the execution happens. For this reason, when the `swapPool` is a timelock it is recommended to use individual approvals, and the `swapPool` should try not to use any NFTs approved for swapping. 
```

Comment:
so this part is a bit confusing to me. Imo it's a con that we have to approve only a subset of the tokens. How do we decide which subset? What if my favorite noun is not included? Then do I have to make a proposal to approve it, and we are back to square one or making a proposal to approve a swap? 

Is the idea here that the attack is
- proposal X requests noun Y to be transferred from the swapPool as part of it's actions.
- proposal X succeeds, but there is a timelock delay before execution
- Bad actor B swaps for noun Y from the swapPool before proposal X executes
- Now proposal Y cannot execute one of it's actions, so it will fail

Is there any other way around this problem? I really think it's a pretty big negative if we can only approve a limited pool. Maybe some way to queue requests that last a certain duration. You can lock a certain amount of ETH to the contract to "reserve" a tokenId. Nobody else can transfer that token for a certain duration. If the `msg.sender` of a call to `swap` is the pool itself (a proposal is executing) then it refunds the locked reservation fee. Otherwise, the reservation expires and you lose your reservation fee. The con is that someone could register all of the tokens but hopefully the fact you have to stake eth and cannot get it back unless the proposal passes would disincentivize anyone from doing this. 

The con here is it adds friction and complexity and confusion to creating a prop that requests a noun

Reply:
If the subset approval is non-negotiable then it would be cool to scheme up some ideas for how to let the community help decide which subset we should approve. For example, I'd love to one day earn enough $nouns to get a frog noun, either via the $noun swap or via the treasury via the native swap mechanism we are creating here. But if the initial batch of approved ids does not include a frog, I'm out of luck. So if we have to go with partial subset approval, it would be cool if I could help select a few frogs so hopefully one would still be there by the time I earn enough $nouns

I think it could be a fun community engagement thing

Reply:
Yea agree `setApprovalForAll` is much better. This is the biggest area of concern IMO.

The attack is exactly as you mentioned, which means it's possible to block any prop which is transferring a Noun if we do approve all.

Yea interesting idea to mitigate this.

> Nobody else can transfer that token for a certain duration
We can't enforce this, since we don't hold the Nouns

One way to get around is to set a swap delay longer than the proposal process (similar to or maybe exactly what you are proposing?), but I think its a really bad UX, and adds complexity in terms of adding a reserveStake + refunding. It also still requires proposers to make sure the Noun they are asking for is not currently being reserved (but with subset approval, we are asking proposers to choose Nouns outside of this subset anyhow).

Open to exploring how can can make `setApprovalForAll` work, without opening up the DAO for a malicious actor blocking the transferring of any Nouns. If we decide subset it the path forward, agree could make it into a fun community exercise to select the treasury Nouns.

Reply:
Or alternatively (and maybe this is what you meant):
* Do `setApprovalForAll`
* Allow instant permissonless swaps all the time
* Add the ability to protect/reserve a tokenId in the `NFTSwapper.sol`, which means noone can `swap` for it
* When someone wants a Noun inside a prop, they need to protect/reverse before they submit their prop

For this to work, I think it needs to cost some (non-refundable) ETH to protect/reserve to prevent someone from locking them all up. It also adds more complexity to implement this, but might be what's needed.